src/lib.rs

```diff
@@ -15,16 +15,27 @@ pub use ser::Serializer;
 
 type Result<T> = std::result::Result<T, Error>;
 
-fn static_str_to_js(s: &'static str) -> JsString {
+// This API isn't great, but there isn't a good way for us to return a reference;
+// it would have to have a static lifetime, and while the cache will live forever,
+// it has to reallocate as it grows, invalidating any references to it.
+//
+// The reason we want to use a reference in the first place is to avoid having to
+// repeatedly clone and destroy these strings in the `JsValue` heap, which takes time.
+fn with_js_string<T, F>(s: &'static str, f: F) -> T
```

Comment:
With this one, because it's a generic function, I'm a bit worried monomorphisation might increase code size significantly, and that's another metric we purposefully tried to keep low (well, as much as possible where the perf hit is small).

Can you check what's the size increase like for Brotli'd Wasm for this change (just compile & check benchmark's size) alongside speed increase?

Reply:
| Benchmark              | `static_str_to_js` | `with_js_string` | Change  |
| ---------------------- | ------------------ | ---------------- | ------- |
| Size (Raw)             | 524KiB             | 588KiB           | +12.21% |
| Size (Brotli)          | 148KiB             | 156KiB           | +5.41%  |
| Parse canada           | 73.45 ops/sec      | 71.76 ops/sec    | -2.29%  |
| Parse citm_catalog     | 184 ops/sec        | 194 ops/sec      | +5.43%  |
| Parse twitter          | 200 ops/sec        | 205 ops/sec      | +2.50%  |
| Serialize canada       | 126 ops/sec        | 123 ops/sec      | -2.38%  |
| Serialize citm_catalog | 206 ops/sec        | 208 ops/sec      | +0.97%  |
| Serialize twitter      | 115 ops/sec        | 115 ops/sec      | +0.00%  |

I didn't realise the improvement was that small; it's not a massive increase in code size either, though.

Reply:
Tbh, looking at numbers, this kind of trade-off doesn't look worth it. (especially keeping in mind that raw Wasm code section increase also somewhat matters, especially for cold startup times)

Reply:
I've changed it to return the string normally again, except that it now returns `&'static JsString` instead of `JsString` using `Box::leak`. This gets the performance benefits without the code size increase:

| Benchmark              | `Clone`       | `Box::leak`   | Change  |
| ---------------------- | ------------- | ------------- | ------- |
| Size (Raw)             | 524KiB        | 524KiB        | +0.00%  |
| Size (Brotli)          | 148KiB        | 148KiB        | +0.00%  |
| Parse canada           | 74.40 ops/sec | 73.57 ops/sec | -1.12%  |
| Parse citm_catalog     | 190 ops/sec   | 204 ops/sec   | +7.37%  |
| Parse twitter          | 210 ops/sec   | 220 ops/sec   | +4.76%  |
| Serialize canada       | 124 ops/sec   | 125 ops/sec   | +0.81%  |
| Serialize citm_catalog | 207 ops/sec   | 222 ops/sec   | +7.25%  |
| Serialize twitter      | 116 ops/sec   | 119 ops/sec   | +2.59%  |

It's still not a huge performance difference, so I'm fine to take it out, but it doesn't hurt now that the code size isn't affected.

Reply:
Yeah if you can just revert this particular optimisation, I'm happy to merge the PR. It should be possible to iterate on JsString opts separately if you still want to.

Reply:
Generally, cloning JsValues is reasonably cheap - pretty much like Rc, except with crossing to JS boundary - so I don't think it's worth spending too much time on it. But it's up to you :)

Reply:
> Yeah if you can just revert this particular optimisation, I'm happy to merge the PR.

Yep, I've already reverted it.